kong/clustering.lua

```diff
@@ -0,0 +1,424 @@
+
+  local c = assert(ws_client:new(WS_OPTS))
+  local uri = "wss://" .. address .. "/v1/outlet?node_id=" ..
+              kong.node.get_id() .. "&node_hostname=" .. utils.get_hostname()
+  local res, err = c:connect(uri, { ssl_verify = true,
+                                    client_cert = CERT,
+                                    client_priv_key = CERT_KEY,
+                                    server_name = "kong_clustering",
+                                  }
+                            )
+  if not res then
+    local delay = 9 + math.random()
+
+    ngx_log(ngx_ERR, "connection to control plane broken: ", err,
+            " retrying after ", delay , " seconds")
+    assert(ngx.timer.at(delay, communicate, conf))
+    return
+  end
+
+  -- connection established
+  -- ping thread
+  ngx.thread.spawn(function()
+    while true do
+      send_ping(c)
+      ngx_sleep(PING_INTERVAL)
+    end
+  end)
+
+  while true do
+    local data, typ, err = c:recv_frame()
+    if err then
+      ngx.log(ngx.ERR, "error while receiving frame from control plane: ", err)
+      c:close()
+
+      local delay = 9 + math.random()
+      assert(ngx.timer.at(delay, communicate, conf))
+      return
+    end
+
+    if typ == "binary" then
+      local msg = assert(cjson_decode(data))
+
+      if msg.type == "reconfigure" then
+        local config_table = assert(msg.config_table)
+
+        local res, err = update_config(config_table, true)
+        if not res then
+          ngx_log(ngx_ERR, "unable to update running config: ", err)
+        end
+
+        send_ping(c)
+
+      end
+    elseif typ == "pong" then
+      ngx_log(ngx_DEBUG, "received PONG frame from control plane")
+    end
+  end
+end
+
+
+function _M.handle_cp_websocket()
+  -- use mutual TLS authentication
+  local cert = ngx_var.ssl_client_raw_cert
+
+  if not cert then
+    ngx_log(ngx_ERR, "Data Plane failed to present client certificate " ..
+                     "during handshake")
+    return ngx_exit(444)
+  end
+
+  cert = assert(openssl_x509.new(cert, "PEM"))
+  local digest = cert:digest("sha256")
+
+  if CERT_DIGEST and digest ~= CERT_DIGEST then
+    ngx_log(ngx_ERR, "Data Plane presented incorrect client certificate " ..
+                     "during handshake, expected digest: " .. CERT_DIGEST ..
+                     " got: " .. digest)
+    return ngx_exit(444)
+  end
+
+  local node_id = ngx_var.arg_node_id
+  if not node_id then
+    ngx_exit(400)
+  end
+
+  local node_hostname = ngx_var.arg_node_hostname
+  local node_ip = ngx_var.remote_addr
+
+  local wb, err = ws_server:new(WS_OPTS)
+  if not wb then
+    ngx_log(ngx_ERR, "failed to perform server side WebSocket handshake: ", err)
+    return ngx_exit(444)
+  end
+
+  local sem = semaphore.new()
+  local queue = { sem = sem, }
+  clients[wb] = queue
+
+  local res
+  -- unconditionally send config update to new clients to
+  -- ensure they have latest version running
+  res, err = declarative.export_config()
```

Comment:
Question for personal understanding: Why do we stored the config in as part of the queue? Why can't we simply have an event for it (dummy table or some such) and then load call export_config at that point?

Also, why don't we cache the current latest config at module level and then use that? There is a call to `export_config` in few places, I think it should be only one place where the config is actually swapped.

Reply:
Good thought. Let me try to dig up the thought process, I think they are as follows:

1. We store the table object in order to not having to call `export_config` multiple times. In practice this is a very cheap operation, as the table itself is not copied in each queue, but all points to the same table instance.
2. Because of the same reason, caching the result of `export_config` at module level is not hugely beneficial, as the `export_config` is only called once for each new version anyway, no matter how much DPs are connected.

Reply:
That makes sense.
The issue I've in mind is that there is a call to `export_config()`  in the worker_event handler as well.
Should there be one place this is called and the reference to the table is kept in one place?
That makes it easier to reason about the flow and we can later on expose the info of what is the current config that is being served by a specific CP node.